core/domain/role_services.py

```diff
@@ -131,3 +163,36 @@ def get_all_actions(role):
         role_actions.extend(get_all_actions(parent_role))
 
     return list(set(role_actions))
+
+
+def get_role_graph_data():
+    """Returns dict for displaying roles graph.
+
+    Returns:
+        dict. A dict containing data in following format:
+        {
+            links: list(dict(str:str)). List of dicts defing each edge in
+                following format:
+                    {
+                        source: RoleId from which edge is going out.
+                        target: RoleId to which edge is incoming.
+                    }
+            nodes: dict(str:str). Mapping of roleId to its human readable
+                format.
+        }
+    """
+    role_graph = {}
+    role_graph['links'] = []
+    role_graph['nodes'] = {}
+    for role in PARENT_ROLES:
+        role_graph['nodes'][role] = get_human_readable_role(role)
+        for parent in PARENT_ROLES[role]:
+            role_graph['links'].append({'source': parent, 'target': role})
+    return role_graph
+
+
+def store_role_query(user_id, intent, method=None, role=None, username=None):
+    """Stores the query to role structure in RoleQueryAuditModel."""
+    audit_models.RoleQueryAuditModel(
+        user_id=user_id, intent=intent,
```

Comment:
Hm, fair point. Let's have one though, so that we at least have some control over how it's generated. Maybe something like `USER_ID.TIMESTAMP_IN_SECS.INTENT.RANDOM_NUMBER'?